allen-cell-animated/nucmorph-colorizer
# fix: Colorizer freezes on missing frame or backdrop

Problem
=======
Closes #143. The colorizer UI used to freeze playback if a frame could not be loaded, which can happen if the user doesn't have network access to the files or if the files have been deleted.

*Estimated review size: small, 10 minutes*

Solution
========
- `ColorizeCanvas` can now handle failed promises, showing a blank frame or backdrop and allowing playback to resume.
- Adds an icon to the canvas when an expected backdrop or frame image is missing.

## Type of change
* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
Missing frames:
1. Open the current build and start playback. The interface should freeze at frame 25/26: 
https://dev-aics-dtp-001.int.allencell.org/nucmorph-colorizer/dist/index.html#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fassay-dev%2Fcomputational%2Fdata%2FEMT_deliverable_processing%2FTranscription_factor_cell_lines%2FVersion_1%2Fcolorizer_output%2FEOMES%2Fcollections_all.json&dataset=AICS-122_3D-MG-EMT-no-MG_fms_id_592af19bfc414ae1bf05b81967190944
3. Open the preview link for this fix and start playback. The interface should continue playing past frame 26: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-284/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fassay-dev%2Fcomputational%2Fdata%2FEMT_deliverable_processing%2FTranscription_factor_cell_lines%2FVersion_1%2Fcolorizer_output%2FEOMES%2Fcollections_all.json&dataset=AICS-122_3D-MG-EMT-no-MG_fms_id_592af19bfc414ae1bf05b81967190944

Missing backdrops:
1. Open the current build and start playback. The interface should freeze at frame 10: https://dev-aics-dtp-001.int.allencell.org/nucmorph-colorizer/dist/index.html#/viewer?dataset=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2F3500005828_25%2Fmanifest.json&keep-range=0&bg-key=ch0ch1ch2
2. Open the preview link for this fix and start playback. The interface should continue playing past frame 10: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-284/#/viewer?dataset=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2F3500005828_25%2Fmanifest.json&keep-range=0&bg-key=ch0ch1ch2

Screenshots (optional):
-----------------------
![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/28a5ae0a-1968-4a72-88ba-d8441ee54b18)

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/bb1aa64f-5269-4dbd-8345-5b830933ee5a)
